test/fft_utest.cpp

```diff
@@ -483,3 +549,244 @@ TYPED_TEST(FftTest, TestFftGf2n) // NOLINT
+    assert(arith::exp<TypeParam>((b - 1), 2) * arith::exp<TypeParam>(p, 2) < m);
+
+    // find x so it is not a quadratic residue in GF_p1 and GF_p2
+    assert(arith::jacobi<TypeParam>(3, p1) == arith::jacobi<TypeParam>(p1, 3));
+    assert(arith::jacobi<TypeParam>(p1, 3) == arith::jacobi<TypeParam>(2, 3));
+    assert(arith::jacobi<TypeParam>(3, p2) == arith::jacobi<TypeParam>(p2, 3));
+    assert(arith::jacobi<TypeParam>(p2, 3) == arith::jacobi<TypeParam>(2, 3));
+    assert(arith::jacobi<TypeParam>(2, 3) == -1);
+    // which means x=3 is not a quadratic residue in GF_p1 and GF_p2
+
+    // therefore we can compute 2^n-th roots of unity in GF_p1 and GF_p2
+    uint64_t w1 = arith::exp<TypeParam>(3, a1);
+    uint64_t w2 = arith::exp<TypeParam>(3, a2);
+    // std::cerr << "w1=" << w1 << " w2=" << w2 << "\n";
+    assert(w1 == 9);
+    assert(w2 == 243);
+
+    // find root of unity in GF_p1p2
+    uint64_t _a[2];
+    uint64_t _n[2];
+    _a[0] = w1;
+    _n[0] = p1;
+    _a[1] = w2;
+    _n[1] = p2;
+    uint64_t w = arith::chinese_remainder<uint64_t>(2, _a, _n);
+    // std::cerr << " w=" << w << "\n";
+    ASSERT_EQ(w, 25559439);
+
+    std::unique_ptr<gf::Field<TypeParam>> gf_m =
+        gf::alloc<gf::Field<TypeParam>, gf::Prime<TypeParam>>(m);
+    fft::Large<TypeParam> fft(*gf_m, l, 25559439);
+
+    // parse the big numbers
+    char X[] = "1236548787985654354598651354984132468";
+    char Y[] = "745211515185321545554545854598651354984132468";
+
+    quadiron::vec::Vector<TypeParam>* _X =
+        this->_convert_string2vec(*gf_m, fft.get_n(), X);
+    // _X->dump();
+    quadiron::vec::Vector<TypeParam>* _Y =
+        this->_convert_string2vec(*gf_m, fft.get_n(), Y);
+    // _Y->dump();
+
+    quadiron::vec::Vector<TypeParam>* sfX =
+        new quadiron::vec::Vector<TypeParam>(*gf_m, fft.get_n());
+    quadiron::vec::Vector<TypeParam>* sfY =
+        new quadiron::vec::Vector<TypeParam>(*gf_m, fft.get_n());
+    quadiron::vec::Vector<TypeParam>* _XY =
+        new quadiron::vec::Vector<TypeParam>(*gf_m, fft.get_n());
+    quadiron::vec::Vector<TypeParam>* sfXY =
+        new quadiron::vec::Vector<TypeParam>(*gf_m, fft.get_n());
+
+    fft.fft(*sfX, *_X);
+    fft.fft(*sfY, *_Y);
+
+    for (int i = 0; i <= fft.get_n() - 1; i++) {
+        quadiron::DoubleSizeVal<TypeParam> val =
+            quadiron::DoubleSizeVal<TypeParam>(sfX->get(i)) * sfY->get(i);
+        _XY->set(i, val % m);
+    }
+
+    fft.ifft(*sfXY, *_XY);
+
+    // carry propagation
+    std::string z("0");
+    for (int i = 0; i <= fft.get_n() - 1; i++) {
+        std::string b(std::to_string(sfXY->get(i)));
+        if (b != std::string("0")) {
+            b.append(i, '0');
+            z = this->big_num_add(z, b);
+        }
+    }
+
+    // std::cout << "z=" << z << "\n";
+    ASSERT_EQ(
+        z,
+        std::string("921490395895362412399910100421159322")
+            + "712298564831565484737491129935640058571771024");
+
+    delete sfXY;
```

Comment:
use `unique_ptr` to avoid these `delete`.

Reply:
+1